packages/grid/x-data-grid/src/colDef/gridSingleSelectColDef.tsx

```diff
@@ -1,15 +1,42 @@
 import { GRID_STRING_COL_DEF } from './gridStringColDef';
-import { GridColTypeDef } from '../models/colDef/gridColDef';
+import { GridColTypeDef, GridColDef } from '../models/colDef/gridColDef';
 import { renderEditSingleSelectCell } from '../components/cell/GridEditSingleSelectCell';
-import {
-  getGridSingleSelectOperators,
-  getGridSingleSelectQuickFilterFn,
-} from './gridSingleSelectOperators';
+import { getGridSingleSelectOperators } from './gridSingleSelectOperators';
+import { getLabelFromValueOption } from '../components/panel/filterPanel/filterPanelUtils';
+
+const isArrayOfObjects = (options: any): options is Array<{ value: any; label: string }> => {
+  return typeof options[0] === 'object';
+};
 
 export const GRID_SINGLE_SELECT_COL_DEF: GridColTypeDef = {
   ...GRID_STRING_COL_DEF,
   type: 'singleSelect',
+  valueFormatter(params) {
+    const { id, field, value, api } = params;
+    const colDef = params.api.getColumn(field);
+
+    let valueOptions: GridColDef['valueOptions'];
```

Comment:
`valueOptions` will never be a function

```suggestion
    let valueOptions: Array<ValueOptions>;
```